storage/innobase/dict/dict0dict.cc

```diff
@@ -7450,27 +7450,130 @@ dberr_t dict_get_dictionary_info_by_id(ulint dict_id, char **name,
   return err;
 }
 
-bool dict_detect_encryption(bool is_upgrade, space_id_t mysql_plugin_space) {
-  bool encrypt_mysql = false;
-  if (is_upgrade) {
-    if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
-      return (srv_sys_space.is_encrypted());
-    }
+/** Detect if database has encrypted mysql plugin space. This is
+ *  indication that system tablespace was encrypted.
+@param[in] mysql_plugin_space   space_id of mysql/plugin table.
+@return true if encrypted, false if not (or fail to open mysql/plugin table) */
+static bool dict_is_mysql_plugin_space_encrypted(
+    space_id_t mysql_plugin_space) {
+  if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
+    return (srv_sys_space.is_encrypted());
+  }
+
+  space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
+  if (space_id == SPACE_UNKNOWN) {
+    return (false);
+  }
 
-    space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
-    if (space_id == SPACE_UNKNOWN) {
-      return (false);
-    }
+  const fil_space_t *space = fil_space_get(space_id);
+  ut_ad(space != nullptr);
 
-    fil_space_t *space = fil_space_get(space_id);
-    ut_ad(space != nullptr);
+  if (space == nullptr) {
+    return false;
+  }
+  return FSP_FLAGS_GET_ENCRYPTION(space->flags);
+}
 
-    if (space == nullptr) {
-      return (false);
+/** Detect if innodb_encrypt_tables is set either to ON or FORCE or
+ONLINE_TO_KEYRING*
+@return true if innodb_encrypt_tables is equal to ON or FORCE or
+ONLINE_TO_KEYRING* */
+static bool dict_should_be_keyring_encrypted() {
+  return srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING_FORCE;
+}
+
+/** Reads mysql.ibd's page0 from buffer
+@return page0 or nullptr */
+static buf_frame_t *get_mysql_ibd_page_0_from_buffer() {
+  fil_space_t *space = fil_space_acquire_silent(dict_sys_t::s_space_id);
+  if (space == nullptr) {
+    return nullptr;
+  }
+
+  const page_size_t page_size(space->flags);
+  mtr_t mtr;
+  mtr_start(&mtr);
+  buf_block_t *block = buf_page_get(page_id_t(dict_sys_t::s_space_id, 0),
+                                    univ_page_size, RW_X_LATCH, &mtr);
+  mtr_commit(&mtr);
+
+  fil_space_release(space);
+
```

Comment:
block should never be nullptr. another debug assert here please

Reply:
actually with PS implementation it can be NULL. If srv_pass_corrupt_tables is ON or there is a decryption error. For MySQL it would just crash.

Reply:
Ok, does it mean all buf_page_get_gen() in PS checks for nullptr?

Reply:
I cannot say for sure all of them, but we should for null when we add new code. In encryption threads branch I have added quite a lot of them - to see if decryption passed.

Reply:
sorry, when decryption fails I am returning error though an [out] variable. nullptr is only for srv_pass_corrupt_tables and it should be check with macro that will assert if the srv_pass_corrupt_tables != NULL. Will update the patch with this assert.

Reply:
OK, I have confused myself for a moment - both for decryption errors and for srv_pass_corrupt_tables the page can be nullptr.